### ISSUE TITLE
per layer loss

### ISSUE THREAD

**Issue by coldgemini** [fizyr/keras-retinanet#261]
can you add loss per fpn layer as facebook detectron is doing. That can be quite helpful when there is a big scale imbalance

**Comment by hgaiser** [comment]
Could you explain more about this difference? Currently there is a loss function for the regression and a loss function for the classification. How does this differ from the loss in Detectron?

**Comment by yhenon** [comment]
As I understand it, it's a loss per scale of the FPN. So you can weigh objects detected on `P6` differently from those from `P5` for example. Not sure how much it helped, but wouldnt be too hard to do.

**Comment by hgaiser** [comment]
> As I understand it, it's a loss per scale of the FPN. So you can weigh objects detected on P6 differently from those from P5 for example. Not sure how much it helped, but wouldnt be too hard to do.

I take it this is not documented in their paper at all? At least I don't remember reading it.

**Comment by awilliamson** [comment]
@hgaiser It is indeed not in the Focal Loss paper, nor any of the Faster/Fast/Mask RCNN papers prior to them, from my recollection. @coldgemini Could you link the academic journal where they outline their success with this?

**Comment by coldgemini** [comment]
I don't think it's another type of loss but it just give additional information on each layer. Some dataset has severe scale imbalance so the per layer loss info can be helpful to determine how many FPN layers to use.

**Comment by hgaiser** [comment]
Can you explain to me how it works in Detectron? I haven't had the chance to look at it yet.

**Comment by hgaiser** [comment]
I recently ran a test where each FPN layer was kept separate and the normalisation was also kept separate. This is what I thought Detectron did, but they normalize globally, not per FPN layer.

Anyway, to make a long story short, basically it was doing a per FPN layer loss, but the implementation looks significantly worse doing that. I'll close this issue as for now I see no reason that justifies that implementation.

**Comment by coldgemini** [comment]
you are rigth, but my point is that still it's helpful to get info for each layer when tuning anchor sizes
